pkg/agent/bgp/gobgp/gobgp.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gobgp
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"time"
+
+	gobgpapi "github.com/osrg/gobgp/v3/api"
+	"github.com/osrg/gobgp/v3/pkg/server"
+	"google.golang.org/protobuf/types/known/anypb"
+	"k8s.io/utils/net"
+
+	"antrea.io/antrea/pkg/agent/bgp"
+)
+
+const (
+	ipv4AllZero = "0.0.0.0"
+	ipv6AllZero = "::"
+)
+
+type Server struct {
+	server       *server.BgpServer
+	globalConfig *gobgpapi.Global
+}
+
+func NewGoBGPServer(globalConfig *bgp.GlobalConfig) bgp.Interface {
+	return &Server{
+		server: server.NewBgpServer(server.LoggerOption(newGoBGPLogger())),
+		globalConfig: &gobgpapi.Global{
+			Asn:        globalConfig.ASN,
+			RouterId:   globalConfig.RouterID,
+			ListenPort: globalConfig.ListenPort,
+		},
+	}
+}
+
+func (s *Server) Start() error {
+	go s.server.Serve()
+	if err := s.server.StartBgp(context.TODO(), &gobgpapi.StartBgpRequest{Global: s.globalConfig}); err != nil {
+		return fmt.Errorf("failed to start BGP server: %w", err)
+	}
+	return nil
+}
+
+func (s *Server) Stop() error {
+	if err := s.server.StopBgp(context.TODO(), &gobgpapi.StopBgpRequest{}); err != nil {
+		return fmt.Errorf("failed to stop BGP server: %w", err)
+	}
+	return nil
+}
+
+func (s *Server) AddPeer(peer bgp.PeerConfig) error {
+	request := &gobgpapi.AddPeerRequest{Peer: toGoBGPPeer(peer)}
+	if err := s.server.AddPeer(context.TODO(), request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) UpdatePeer(peer bgp.PeerConfig) error {
+	request := &gobgpapi.UpdatePeerRequest{Peer: toGoBGPPeer(peer)}
+	if _, err := s.server.UpdatePeer(context.TODO(), request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) RemovePeer(peer bgp.PeerConfig) error {
+	request := &gobgpapi.DeletePeerRequest{Address: peer.Address}
+	if err := s.server.DeletePeer(context.TODO(), request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) GetPeers() ([]bgp.PeerStatus, error) {
+	var peerStatuses []bgp.PeerStatus
+
+	fn := func(peer *gobgpapi.Peer) {
+		if peer == nil {
+			return
+		}
+		peerStatus := bgp.PeerStatus{}
+		if peer.Transport != nil {
+			peerStatus.Port = int32(peer.GetTransport().GetRemotePort())
```

Comment:
I don't think you should mix "styles" for accessing fields, e.g., `peer.Transport` vs `peer.GetTransport`

You may be able to use something like this:
```go
if transport := peer.GetTransport(); transport != nil {

}
```

Reply:
Another note: I don't know in which conditions these fields (e.g., `Transport`) can be `nil`, but what is the behavior we want in this case?

Edit: looking at the gobgp code, it seems that these fields are always populated when listing peers. See https://github.com/osrg/gobgp/blob/0148e2d22dcf40655d2b75c6f40bf49a378c4740/pkg/config/oc/util.go#L437. The fact that they are pointers is probably for configuration only. We can keep the `nil` checks but we should add a comment.

Reply:
> I don't think you should mix "styles" for accessing fields

Thanks for pointing out this

> Edit: looking at the gobgp code, it seems that these fields are always populated when listing peers. See https://github.com/osrg/gobgp/blob/0148e2d22dcf40655d2b75c6f40bf49a378c4740/pkg/config/oc/util.go#L437. The fact that they are pointers is probably for configuration only. We can keep the nil checks but we should add a comment.

You are right. All pointer fields are not nil when peers are listed. If so, I think we could remove the nil checks, but it seems that it is harmless and safe to reserve them. Not sure which is better.